src/scikit_build_core/program_search.py

```diff
@@ -114,11 +104,19 @@ def get_cmake_program(cmake_path: Path) -> Program:
 def get_cmake_programs(*, module: bool = True) -> Generator[Program, None, None]:
     """
     Get the path and version for CMake. If the version cannot be determined,
-    yiels (path, None). Otherwise, yields (path, version). Best matches are
-    yielded first.
+    yields (path, None). Otherwise, yields (path, version). Best matches are
+    yielded first; module, then cmake and cmake3 sorted by highest version first.
     """
-    for cmake_path in _get_cmake_path(module=module):
-        yield get_cmake_program(cmake_path)
+    if module:
+        with contextlib.suppress(ImportError):
+            # If a "cmake" directory exists, this will also ImportError
+            from cmake import CMAKE_BIN_DIR
+
+            yield get_cmake_program(Path(CMAKE_BIN_DIR) / "cmake")
```

Comment:
Would need to patch downstream CMake dummy file better if something like this is used:
https://src.fedoraproject.org/rpms/cmake/blob/rawhide/f/cmake.spec#_353-366

My thought is to separate that into 2 packages. The cmake-pypi build that is running `scikit-build-core` and the minimal wrapper that contains the `cmake` package. That way the latter can be easier to synchronize across pypi and distros, and the former can contain more information on the packaging vendor (pypi vs Fedora etc.)

Reply:
Fedora should not be providing a dist-info file for `cmake`. `"cmake"` (and `"ninja"`) are PyPI redistributions. I believe this still work, though, as importing CMAKE_BIN_DIR should trigger the ImportError.

Reply:
> Fedora should not be providing a dist-info file for `cmake`

The issue we've had there is that some projects add `cmake` to the `build-system.requires`, sometimes with version specification. Ideally adding that should trigger a warning since the `cmake` and `ninja` version specifications can now be handled within `tools.scikit-build`, and all of those packages should be changed over.

The dist-info file was added because that was the simplest way to satisfy the `build-system.requires` without much patching upstream.

So far these are only in `rawhide`, but they are being picked up for RHEL10 fork soon, so it might be good to discuss this soon. Most relevant python devs are on [fedora-python](https://matrix.to/#/#python:fedoraproject.org) matrix channel if you have access to this, or I can try to ping some other devs to chime in either on Github or Discord?

Reply:
This is largely my fault. I assumed cmake and ninja PyPI projects only provide the executables and I haven't checked (I should have checked).

Reply:
More like my fault because I knew about those files, but did not expect them to be used and was considering how to add them. We have a few points to discuss:
- What to do with the current dummy package in rawhide? We could resolve it with `Obsoletes`, but we might need to be a bit quick about it so that it's not affecting RHEL10
- Is it possible to incorporate the python files? Yes, it should be but there are 2 issues:
  - The pypi releases may lag behind or lack some pre-releases that are packaged for Fedora
  - How to distinguish between which `vendor` of CMake is being used
- Would other tools also use the pip registry/python modules? Would there be situations where it would be useful to still have these files on the distro side?

Reply:
RHEL10 is no concern, it was forked form Fedora way before we added that packages.

Reply:
@henryiii I am pinging on this discussion because @hroncok is also present here. What do we do for projects that have `build-system.requires` with `cmake`, but do not use `scikit-build-core`, e.g.` setuptools`, `scikit-build`, `poetry-core` [^1]?

[^1]: https://sourcegraph.com/search?q=context:global+file:pyproject%5C.toml%24+requires.*cmake&patternType=regexp&sm=0